egg/core/losses.py

```diff
@@ -0,0 +1,108 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class DiscriminationLoss:
+    def __call__(self, sender_input, _message, _receiver_input, receiver_output, labels):
+        return self.discrimination_loss(receiver_output, labels)
+
+    @staticmethod
+    def discrimination_loss(receiver_output, labels):
+        acc = (receiver_output.argmax(dim=1) == labels).detach().float()
+        loss = F.cross_entropy(receiver_output, labels, reduction="none")
+        return loss, {'acc': acc}
+
+
+class RecoLoss:
+    def __init__(self, n_attributes: int, n_values: int, batch_size: int):
+        self.n_attributes = n_attributes
+        self.n_values = n_values
+        self.batch_size = batch_size
+
+    def __call__(self, sender_input, _message, _receiver_input, receiver_output, labels):
+        self.reconstruction_loss(
+            receiver_output,
+            labels,
+            self.batch_size,
+            self.n_attributes,
+            self.n_values
+        )
+
+    @staticmethod
+    def reconstruction_loss(
+        receiver_output,
+        labels,
+        batch_size,
+        n_attributes,
+        n_values
+    ):
+        receiver_output = receiver_output.view(batch_size * n_attributes, n_values)
+        receiver_guesses = receiver_output.argmax(dim=1)
+        correct_samples = (receiver_guesses == labels.view(-1)).view(batch_size, n_attributes).detach()
+        acc = (torch.sum(correct_samples, dim=-1) == n_attributes).float()
+        labels = labels.view(batch_size * n_attributes)
+        loss = F.cross_entropy(receiver_output, labels, reduction="none")
+        loss = loss.view(batch_size, -1).mean(dim=1)
+        return loss, {'acc': acc}
+
+
+class NTXentLoss:
```

Comment:
Unlike other losses, this thing is huuge. Do you think it will make sense to add a short description/py doc?

Reply:
yep, I would like to add some doctest as well, especially if we rewrite some games using these